cabot/cabotapp/models.py

```diff
@@ -633,7 +649,13 @@ def save(self, *args, **kwargs):
             if get_success_with_retries(recent_results, self.retries):
                 self.calculated_status = Service.CALCULATED_PASSING_STATUS
             else:
-                self.calculated_status = Service.CALCULATED_FAILING_STATUS
+                last_result = recent_results[0] if recent_results and len(recent_results) > 0 else None
+                if last_result and last_result.acked:
```

Comment:
Yeah, you're right!

An easy fix is to retroactively mark the latest result as `acked=True` (if it matches the acks filters), but I don't really like rewriting history. This might be okay, since we're not actually changing the result's data, just whether or not it's acked.

Another solution is to immediately run the associated check and update associated services when we create an ack. We can schedule the updates as celery tasks, but the downside is they won't finish before the ack create request returns, so the check/service won't show as 'acked' until the user refreshes the page. If we do the updates inside the create ack request (like when you click the manual 'run check' button), it should also show that the service as acked immediately...but if the check takes a long time to run, the create ack request could time out. This is honest, but might have a worse UX.

A third option is to create a new dummy result, with the same status/tags as the latest result, but with `acked` set to True (again, if the filters match) and the message 'ack created'. This is like (1), but we don't mutate existing results. It's still a little dishonest, but the message should make it clear what happened.

Reply:
Can we change the source of truth on whether the check is acked to be on the check itself (update this directly when the check is acked and when the ack is deleted/expired) and then just have the result values as a log of what happened (can determine their values based on the check)? All the result-based solutions seem a little sketch/dishonest like you mentioned. 

Only problem then is determining the service status after its ack expires--we could maybe calculate a shadow result for every acked result and use that for the service status if the check isn't acked? Or we could not have acks on the status check results at all, but I do like showing the values as a log of what happened...

Reply:
The source of truth of whether the check is acked *is* the check's status (through `check.calculated_status`), but the status check results give a log for when/why the check had each status.

We could just set `check.calculated_status = ACKED` when we create an ack...the next time the check runs, it'll update calculated_status properly (so if the ack got closed in the mean time the check will go back to normal). We wouldn't have a log on the check's page for the status change though.

This still sounds a bit sketch, tbh... I think we should just queue up check.run() with celery, and have the server poll for task completion for up to ~5 seconds. If the task doesn't finish in that time, we can flash a message to the user on page load explaining that the ack will take effect once the check finishes running.

---

Round-up of the options to solve the delay in checks + services getting "acked" status after creating an ack:

1. On ack creation, set `check.calculated_status = ACKED` if the latest result matches the ack's filter. Immediately update any associated services.

Pros:
* Immediate
* Doesn't change results

Cons:
* No log on the check page of when/why the status became acked

2. On ack creation, retroactively mark the last status check result as `acked = True` (if filters match) and set `check.calculated_status = ACKED`. Immediately update any associated services.

Pros:
* Immediate

Cons:
* Changes history

3. On ack creation, create a new acked result based on the latest result:

```python
last_result = StatusCheckResult.objects.filter(status_check=self).order_by('-id').first()
if last_result and ack.matches(last_result):
    new_result = StatusCheckResult(status_check=check, succeeded=False, error="ack created", acked=True)
    new_result.save()
    new_result.tags.add(*last_result.tags)
    check.last_run = new_result
    check.save()  # updates check status based on recent results
```

Immediately update any associated services.

Pros:
* Immediate
* Logs when/why the status became acked

Cons:
* Might interact with the "success with retries" system
* Still a little bit dishonest (the check didn't *actually* run)

4. After ack creation, call `check.run()` and `[s.update_status() for s in check.services]` before rendering the page for the user.

Pros:
* "Immediate," depending on how long the check runs
* Honest
* Simple to implement

Cons:
* If the check takes too long to run, the web request will time out with a 504.
* Page appears to hang while the check is running.

5. After ack creation, schedule `check.run()` and service updates with celery. Poll for task completion for up to X seconds to prevent a timeout/504. Flash a warning to the user if the task doesn't complete within the timeout.

Pros:
* Immediate
* Honest

Cons:
* Check/service may not be updated immediately for checks that take a long time to run, but the user will at least know what's going on
* Not sure how to do polling with celery (might be tricky since we set `ignore_result`? may just need to do a new task)
* Page may still hang for up to X seconds

Reply:
I think I'm good with 4 or 5...we already do run checks synchronously in web requests when you manually run a check, so it wouldn't be that crazy to do it here.

The other thing with this is that a check isn't really acked for the amount of time specified. Example: check A runs every 45 minutes
- 10:00: ACK A for 2 hours, A runs and is acked
- 10:45: A runs and is acked
- 11:30: A runs and is acked
- 12:15: A runs and is no longer acked
So it's 2 hours 15 minutes instead of 2 hours. This doesn't seem unreasonable given that the check status only updates at the check interval anyways...but just wanted to point it out

Reply:
Great point - we should apply the same solution whenever an ack is closed too (i.e. immediately schedule the check to run).

I'd like to give (5) a try...if it doesn't work, we can always fall back to (4).

Reply:
Alright, implemented (5). Ended up polling the DB to check when the check finishes running, since we don't currently configure a celery result store.